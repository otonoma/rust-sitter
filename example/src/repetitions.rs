pub mod grammar {
    use rust_sitter::{Rule, Spanned};

    #[derive(Debug, Rule)]
    #[language]
    #[allow(dead_code)]
    pub struct NumberList {
        #[sep_by1(",")]
        #[leaf(pattern(r"\d+"))]
        numbers: Spanned<Vec<Spanned<i32>>>,
    }

    #[derive(Rule)]
    #[extra]
    struct Whitespace {
        #[leaf(pattern(r"\s"))]
        _whitespace: (),
    }
}

// TODO: Currently not allowed, needs to be fixed.
// pub mod grammar2 {
//     use rust_sitter::{Rule, Spanned};
// 
//     #[derive(Debug, Rule)]
//     #[language]
//     #[allow(dead_code)]
//     pub struct NumberList {
//         #[leaf(pattern(r"\d+"))]
//         numbers: Spanned<Vec<Spanned<i32>>>,
//     }
// 
//     #[derive(Rule)]
//     #[extra]
//     struct Whitespace {
//         #[leaf(pattern(r"\s"))]
//         _whitespace: (),
//     }
// }
// 
// pub mod grammar3 {
//     use rust_sitter::{Rule, Spanned};
// 
//     #[derive(Debug, Rule)]
//     #[language]
//     #[allow(dead_code)]
//     pub struct NumberList {
//         #[sep_by(",")]
//         #[leaf(pattern(r"\d+"))]
//         numbers: Spanned<Vec<Spanned<Option<i32>>>>,
//         #[skip(123)]
//         metadata: u32,
//     }
// 
//     #[derive(Rule)]
//     #[extra]
//     struct Whitespace {
//         #[leaf(pattern(r"\s"))]
//         _whitespace: (),
//     }
// }

#[cfg(test)]
mod tests {
    use super::*;
    use rust_sitter::Language;

    #[test]
    fn repetitions_grammar() {
        // Bug in latest tree-sitter: empty parse on a top-level repeat1 segfaults.
        insta::assert_debug_snapshot!(grammar::NumberList::parse(""));
        insta::assert_debug_snapshot!(grammar::NumberList::parse("1"));
        insta::assert_debug_snapshot!(grammar::NumberList::parse("1, 2"));
    }

    // #[test]
    // fn repetitions_grammar2() {
    //     insta::assert_debug_snapshot!(grammar2::parse(""));
    //     insta::assert_debug_snapshot!(grammar2::parse("1"));
    //     insta::assert_debug_snapshot!(grammar2::parse("1 2"));
    // }

    // #[test]
    // fn repetitions_grammar3() {
    //     insta::assert_debug_snapshot!(grammar3::parse(""));
    //     insta::assert_debug_snapshot!(grammar3::parse("1,"));
    //     insta::assert_debug_snapshot!(grammar3::parse("1, 2"));
    //     insta::assert_debug_snapshot!(grammar3::parse("1,, 2"));
    //     insta::assert_debug_snapshot!(grammar3::parse("1,, 2,"));
    // }
}
