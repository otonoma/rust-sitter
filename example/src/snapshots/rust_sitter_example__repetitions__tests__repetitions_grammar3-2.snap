---
source: example/src/repetitions.rs
expression: "grammar3::parse(\"1,\")"
---
Ok(
    NumberList {
        numbers: Spanned {
            value: [
                Spanned {
                    value: Some(
                        1,
                    ),
                    byte_span: (
                        0,
                        1,
                    ),
                    line_span: (
                        Point {
                            line: 1,
                            column: 1,
                        },
                        Point {
                            line: 1,
                            column: 2,
                        },
                    ),
                },
            ],
            byte_span: (
                0,
                2,
            ),
            line_span: (
                Point {
                    line: 1,
                    column: 1,
                },
                Point {
                    line: 1,
                    column: 3,
                },
            ),
        },
        metadata: 123,
    },
)
